x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/allocation/TrainedModelAllocationClusterService.java

```diff
@@ -0,0 +1,376 @@
+            public ClusterState execute(ClusterState currentState) {
+                return createModelAllocation(currentState, params);
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                listener.onResponse(TrainedModelAllocationMetadata.metadata(newState).getModelAllocation(params.getModelId()));
+            }
+        });
+    }
+
+    public void removeModelAllocation(String modelId, ActionListener<AcknowledgedResponse> listener) {
+        clusterService.submitStateUpdateTask("delete model allocation", new ClusterStateUpdateTask() {
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                return removeAllocation(currentState, modelId);
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                listener.onResponse(AcknowledgedResponse.TRUE);
+            }
+        });
+    }
+
+    private static ClusterState update(ClusterState currentState, TrainedModelAllocationMetadata.Builder modelAllocations) {
+        if (modelAllocations.isChanged()) {
+            return ClusterState.builder(currentState)
+                .metadata(
+                    Metadata.builder(currentState.metadata()).putCustom(TrainedModelAllocationMetadata.NAME, modelAllocations.build())
+                )
+                .build();
+        } else {
+            return currentState;
+        }
+    }
+
+    ClusterState createModelAllocation(ClusterState currentState, StartTrainedModelDeploymentAction.TaskParams params) {
+        TrainedModelAllocationMetadata.Builder builder = TrainedModelAllocationMetadata.builder(currentState);
+        if (builder.hasModel(params.getModelId())) {
+            throw new ResourceAlreadyExistsException("allocation for model with id [" + params.getModelId() + "] already exist");
+        }
+        builder.addNewAllocation(params);
+        // TODO here is where to handle allocation options if possible
+        TrainedModelAllocation.Builder allocationBuilder = TrainedModelAllocation.Builder.empty(params);
+        for (DiscoveryNode node : currentState.getNodes().getAllNodes()) {
+            if (allocationBuilder.canAllocateToNode(node) && isNodeShuttingDown(currentState, node.getId()) == false) {
+                Optional<String> maybeError = nodeHasCapacity(currentState, params, node);
+                if (maybeError.isPresent()) {
+                    try {
+                        builder.addFailedNode(params.getModelId(), node.getId(), maybeError.get());
```

Comment:
adding a failed node here allows the top level caller see why a particular model is not loaded on a node on which is SHOULD be loaded.